### ISSUE TITLE
Should we cache optimal abilities until the algorithm can be made faster?

### ISSUE THREAD

**Issue by EZtouch** [NightShadeI/Revolution-Rotation-Calculator-#11]
# Algorithm:

_User inputs their config settings.
User inputs set of abilities._
If user config settings **and** abilities have been used by us before
    Print optimal bar to the user, saving hours/days of their time.

Ideally, there would be another app, Revolution Rotation Browser, which shows abilities with the highest AADPT (Average Ability Damage Per Tick), where both apps communicate with one another, allowing submitting ability bar from Revolution Rotation Calculator to the Revolution Calculation Browser.

**Comment by NightShadeI** [comment]
I really love this idea, however won't this require some place to store these bars and over time without some good search method, won't it take a while to search through these bars? (which have been used before).

**Comment by EZtouch** [comment]
Here's what I'm thinking
First, loading cached data should be optional, requiring a user prompt.
When the algorithm starts calculating the optimal bar, call an asynchronous function (LoadJson).
The LoadJson function loads our predefined set of bars/settings, comparing them with the user's abilities and configuration until a valid match is found, if no match is found, the function returns nothing and the algorithm continues as it is. However, in case a match is found, halt every operation, print the cached bar, and show the press enter to exit prompt.

_Apparently, ujson library is pretty fast according to some benchmarks._

Some issues that I can think of in regards to this:
If a value or algorithm change occurs in the future, it would invalidate every cached entry. We need to update this file with every breaking change, which would take a very long time.

Another option is to create an online spreadsheet where users submit their data and give a hyperlink to the page. The problem with this approach is that we cannot validate the user entries or track the version of Revolution Rotation Calculator they're using, since some values did change in the past, which produced different optimal rotations.

I've logged the output I got in the following table, as you can see, in most cases, the higher the number of slots used, the higher the average ability damage per tick is.


Time Ran (s) | Ability Damage % | AADPT % | Slot 1 | Slot 2 | Slot 3 | Slot 4 | Slot 5 | Slot 6 | Slot 7 | Slot 8 | Slot 9 | Slot 10 | Slot 11 | Slot 12 | Slot 13 | Slot 14
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
300 | 22,747.6 | 45.495 | Hurricane | Cleave | Dismember | Sever | Slaughter | Smash | Assault | Quake | Stomp | Slice | Fury |   |   |  
300 | 22,658.6 | 45.317 | Pulverise | Hurricane | Dismember | Cleave | Smash | Assault | Sever | Slice | Quake | Fury |   |   |   |  
300 | 22,540.9 | 45.082 | Hurricane | Cleave | Slaughter | Sever | Quake | Dismember | Assault | Smash | Slice | Fury |   |   |   |  
300 | 22,037.9 | 44.076 | Assault | Hurricane | Cleave | Dismember | Sever | Smash | Slaughter | Fury | Slice |   |   |   |   |  
300 | 20,657.0 | 41.314 | Assault | Cleave | Pulverise | Dismember | Slaughter | Sever | Smash | Slice | Fury |   |   |   |   |  
300 | 19,583.8 | 39.168 | Assault | Slaughter | Cleave | Dismember | Sever | Smash | Slice | Fury |   |   |   |   |   |  
300 | 19,061.1 | 38.122 | Hurricane | Cleave | Smash | Sever | Dismember | Fury | Slice |   |   |   |   |   |   |  
300 | 18,993.7 | 37.987 | Assault | Cleave | Dismember | Sever | Smash | Slice | Fury |   |   |   |   |   |   |  
300 | 18,895.5 | 37.791 | Cleave | Overpower | Dismember | Pulberise | Smash | Sever | Forceful   Backhand | Fury | Slice |   |   |   |   |  
300 | 17,577.0 | 35.154 | Smash | Sever | Dismember | Cleave | Fury | Overpower | Slice |   |   |   |   |   |   |  
300 | 17,577.0 | 35.154 | Smash | Sever | Dismember | Cleave | Fury | Pulberise | Slice |   |   |   |   |   |   |  
300 | 17,086.6 | 34.173 | Quake | Cleave | Sever | Dismember | Smash | Fury | Slice |   |   |   |   |   |   |  
300 | 17,076.3 | 34.153 | Cleave | Forceful   Backhand | Smash | Dismember | Sever | Fury | Slice |   |   |   |   |   |   |  
300 | 17,076.3 | 34.153 | Cleave | Stomp | Smash | Dismember | Sever | Fury | Slice |   |   |   |   |   |   |  
300 | 17,019.5 | 34.039 | Cleave | Sever | Smash | Dismember | Slaughter | Fury | Slice |   |   |   |   |   |   |  
300 | 16,388.8 | 32.778 | Cleave | Smash | Sever | Dismember | Fury | Slice |   |   |   |   |   |   |   |

**Comment by NightShadeI** [comment]
Sure, I am happy with this :) . Just a few things we still need to consider:
- We need to somehow store potentially lots of different bars and settings especially if this project ends up getting big.
- When changes are made to ability values (by Jagex) each configuration recorded with these abilities would need to be removed since there could be potentially new optimal bars now.
- To make the search efficient (since a linear search may have to potentially go through thousands of generated bars) we need to define some way to sort the configurations so that a search (such as binary may be used) to check if any configurations exists.

**Comment by NightShadeI** [comment]
Should we start this up under the project section? sounds fun 👍

**Comment by Ditiae** [comment]
Should we make like a discord? For more generalized conversation on the projectt? Would allow for those unfamiliar with GitHub or just want to make suggestions and stuff to be able to take part and talk about it.

**Comment by EZtouch** [comment]
> Should we start this up under the project section? sounds fun 👍

Sure, you may attempt it, however I will most probably be away for around half a day. 🙂



> Should we make like a discord? For more generalized conversation on the project? Would allow for those unfamiliar with GitHub or just want to make suggestions and stuff to be able to take part and talk about it.

If we're starting a discord channel, GitHub will have to remain the main hub for our discussions, as it is more organized and everyone can check the status of the project and whatnot. However, I believe having user-related discussions on the RuneScape subreddit is more appropriate than Discord.

**Comment by Ditiae** [comment]
> If we're starting a discord channel, GitHub will have to remain the main hub for our discussions, as it is more organized and everyone can check the status of the project and whatnot. However, I believe having user-related discussions on the RuneScape subreddit is more appropriate than Discord.

Fair enough, just figured more people would be willing to join and keep up with discussions regarding this there rather than going here or finding posts on the subreddit. But I get what you mean.

**Comment by MatthewMawby** [comment]
> We need to somehow store potentially lots of different bars and settings especially if this project ends up getting big.
* If the settings/bars are stored as JSON they could easily be saved in a noSQL database such as mongo or dynamo.
> When changes are made to ability values (by Jagex) each configuration recorded with these abilities would need to be removed since there could be potentially new optimal bars now.
* A version number could be saved with each ability/configuration option in the JSON. When the version number is different the document hash will also change, this will prevent new configurations from retrieving ability bars for old configurations with different versions of abilities/gear values. Then a linear scan of all documents can be performed in the background to remove outdated configurations and ability bars since we know the latest version numbers. 
> To make the search efficient (since a linear search may have to potentially go through thousands of generated bars) we need to define some way to sort the configurations so that a search (such as binary may be used) to check if any configurations exists.
* This can be accomplished by hashing the JSON file containing the configuration settings, then saving the documents to a database using the hash as the primary key.

**Comment by NightShadeI** [comment]
You got some very nice thoughts here! 👍 to be honest, I don't have much experience with databases so if you would be able to help here that would be massively appreciated by me and many others.

**Comment by MatthewMawby** [comment]
My workload is a bit high right now, but I could help get that set up when the time comes for it. There's also a good chance that the setting documents could be saved locally to the filesystem with the filename being the hash of the file since there would likely only need to be one instance of the server, and thus no need to retrieve the documents from multiple hosts. The same logic could be applied to periodically scan the directory where the files are saved and invalidate outdated ones and quick retrieval could be done just by hashing the new settings document and determining if there is a matching file.
